Add Tensor.multinomial

marcellofuschi opened the pull request:
Only with replacement (for now)

geohot commented:
Locked the bounty for you

marcellofuschi commented:
Fixed the first assert — the function now works with both 1D and 2D inputs. You cannot pass 3D or higher, just like torch

As for the second assert, sampling without replacement is trickier to implement lazily, so I'll need to think about it. Meanwhile, I'll set the PR as ready in case it's already useful just with replacement.

geohot commented on tinygrad/tensor.py:
```diff
@@ -216,6 +216,19 @@ def kaiming_normal(*shape, a:float = 0.01, **kwargs) -> Tensor:
+  @staticmethod
+  def multinomial(p: Tensor, num_samples: int, replacement: bool = False) -> Tensor:
+    assert p.ndim <= 2, "p must be 1 or 2 dim"
+    assert replacement, "supported only with replacement"
```
If num_samples is 1, is replacement=false okay?

marcellofuschi replied:
yep, I'll allow it

imaolo commented:
Does any runtime checking need to happen (looking at torch.multinomial with negative values). I don't think there are OPs that raise exceptions, atleast I don't know about them.

geohot commented on test/test_randomness.py:
```diff
@@ -96,6 +96,22 @@ def test_kaiming_normal(self):
+    _check_with_torch(p=[[0.453, 0., 1., 0.81], [0.1, 0.8, 0., 0.1]], num_samples=1000, replacement=True)
+    # without-replacement isn't supported, unless taking only one sample
+    self.assertRaises(AssertionError, lambda: Tensor.multinomial(Tensor([0.5, 0.5]), num_samples=100, replacement=False))
+    _check_with_torch(p=[[0.5, 0.5], [0.99, 0.01]], num_samples=1, replacement=False)
```
Umm, doesn't this just fail 1% of the time?

marcellofuschi replied:
It never fails actually, because for `num_samples=1` the ks test isn't reliable. That line was just checking that no error was raised when passing `num_samples=1, replacement=False`.

I've just changed the test code to make it more reliable by accumulating the results of repeated experiments.

Regarding the reliability of `equal_distribution` in general, I wrote a counterexample to test that the samples are considered from different distributions when they should. I ran it 1000 times and it always passed. I don't know if we should keep it in the code since the other tests in the file have no such thing.

geohot replied:
Counterexample is great! Confirms the tests are actually testing something

geohot commented on tinygrad/tensor.py:
```diff
@@ -216,6 +216,19 @@ def kaiming_normal(*shape, a:float = 0.01, **kwargs) -> Tensor:
     std = math.sqrt(2.0 / (1 + a ** 2)) / math.sqrt(prod(shape[1:]))
     return Tensor.normal(*shape, mean=0.0, std=std, **kwargs)
 
+  @staticmethod
```
This shouldn't be a staticmethod, and the first arg should be self.

Fixes `p.multinomial` and is a line shorter

marcellofuschi commented:
Ok, I removed staticmethod and changed the tests accordingly.

Plus, I found a bug with an edge case (input 2D tensor with only 1 row should still return 2D). I added an assert and fixed it.

geohot commented:
Sweet, can pay bounty via PayPal or USDC. E-mail george@tinygrad.org (or DM on discord) with your info to claim!
